### PR TITLE
unityhub: 2.3.2 -> 3.4.1

### LINKED ISSUE #176783: UnityHub new 3.0 package change
**Project description**
Unity is changing the way they deliver the editor via the UnityHub.  They have shifted away from the appimage and have started Debian and RHEL repositories.  They are not producing the appimage anymore.

Would it be possible to add the new UnityHub to NixOS?

The new download instructions are here:
https://docs.unity3d.com/hub/manual/InstallHub.html#install-hub-linux

### PR DESCRIPTION
**Pull request by huantianad**
###### Description of changes
Resolves #176783.

Major version bump of unity hub, switches from appimage installation to new deb installation.

Main thing I'm not sure on is the difference between targetPkgs and multiPkgs in this situation. Which should I use for all these libraries?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by anilanar** [comment]
I ran the app, logged in. At some point, I wanted to change the project directory, so I clicked a button that was going to open a file explorer. It crashed. Now I cannot open `unityhub` anymore, it crashes with a gpu related error:

```
[162555:1227/174555.154527:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162555:1227/174555.354212:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162555:1227/174555.536978:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162555:1227/174555.680980:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162555:1227/174555.804655:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162555:1227/174555.934669:ERROR:gpu_process_host.cc(974)] GPU process exited unexpectedly: exit_code=256
[162729:1227/174555.939717:ERROR:gpu_init.cc(481)] Passthrough is not supported, GL is disabled, ANGLE is 
```

Not sure if this is reproducible by others or a problem on my end.

**Comment by anilanar** [comment]
Temporary workaround:  

```
unityhub --disable-gpu 
```

**Comment by huantianad** [comment]
Hm can't seem to reproduce it on my end, might be an upstream electron error. I'll do some more research.

**Comment by anilanar** [comment]
@huantianad I gave up as unityhub was freezing the whole xserver with simple demo projects. Not sure if it's related to either nvidia drivers or to x11 (versus wayland).

**Comment by huantianad** [comment]
Huh I haven't experienced that myself on Nvidia and X11 myself. There was a recent bug with the editor that could cause it to crash on Linux, but probably unrelated.

**Comment by anilanar** [comment]
Could be also related to me using a different kernel and the beta nvidia driver:

```
  boot.kernelPackages = pkgs.linuxPackages_xanmod;
  hardware.nvidia.package = config.boot.kernelPackages.nvidiaPackages.beta;
```

**Comment by headblockhead** [comment]
I don't know if this might help, and it is not strictly related to UnityHub, but I personally have had the best experience using Unity 2021.2.17f1, as other versions fail to launch, or can crash when the Play button is pressed.

**Comment by huantianad** [comment]
@headblockhead is this the same error as mentioned in [this unity bug report](https://issuetracker.unity3d.com/issues/linux-editor-crashes-at-burst-signal-handler-when-opening-floating-windows)?

**Comment by headblockhead** [comment]
I don't believe so, as I have done some testing, and Unity LTS 2021.3.16f1 works, but Unity 2022.2.2f1 has some weird errors and graphical issues rendering the Universal Render Pipeline example project - as seen in the images.

Screenshot from 2023-01-19 19-43-29

Screenshot from 2023-01-19 19-42-41

Both were marked as fixed for the issue you sent me, and the symptoms don't match what I experienced before, so I don't believe that that issue is related.

File: pkgs/development/tools/unityhub/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/unityhub/default.nix:36]:
```suggestion
```

File: pkgs/development/tools/unityhub/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/unityhub/default.nix:85]: Why not use unpackCmd?

File: pkgs/development/tools/unityhub/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/unityhub/default.nix:97]: We should probably remove the unused wrapper

File: pkgs/development/tools/unityhub/default.nix
Comment (huantianad) [reply pkgs/development/tools/unityhub/default.nix:85]: What's the proper way to do this? Searching GitHub for "repo:NixOS/nixpkgs unpackCmd dpkg" reveals a bunch of ways to do this. Originally I was considering switching to `unpackPhase` with `dpkg-deb -x $src .`

File: pkgs/development/tools/unityhub/default.nix
Comment (huantianad) [reply pkgs/development/tools/unityhub/default.nix:97]: `makeWrapper` should overwrite the existing wrapper, but it still might be worth removing it outright first for clarity.
```suggestion
    # `unityhub` is a shell wrapper that runs `unityhub-bin`
    # Which we don't need and replace with our own custom wrapper
    rm$out/opt/unityhub/unityhub
```

**Comment by huantianad** [comment]
@headblockhead I see, I personally don't use unity 2022, only 2021 so I haven't encountered this issue.

**Comment by huantianad** [comment]
I've made the requested changes, as well as split the packages into `targetPkgs` and `multiPkgs`, which I'm not sure is necessary.

**Comment by huantianad** [comment]
Gonna @SuperSandro2000 for further review

**Comment by hyphenrf** [comment]
I should mention that I tried this, and had to disable firewall for signup to work

**Comment by huantianad** [comment]
Interesting, I didn't sign up in the hub itself, just on their website, so I didn't notice anything. Is there anything we could do about that?

**Comment by hyphenrf** [comment]
Yeah, similar experience here.. The signup process is basically:
- open the hub
- click sign in, it opens the default browser window (likely with xdg-open)
- sign in on the browser, it redirects to a `unityhub://login?...` URL <-- firefox by default (?) prevents this from going through
- if x-scheme-handler is set up correctly, and the browser allows redirection, the hub picks up the URL <-- possible failure point
- once the token is picked up from the URL, finalize the sign-in process in the hub <-- fails here because of firewall (?)

I'm not sure how much of this process are we supposed to streamline from our side.

**Comment by Philipp-M** [comment]
Well it's obviously a NixOS issue (thus the package may already be fine as is, I haven't tested it yet, maybe I'll come to this later).
So I guess the best way to fix it, is to provide a NixOS option `programs.unityhub.enable` that does all the relevant things necessary to run it on NixOS without issues (i.e. set xdg-mime and maybe open the relevant port(s) on the firewall, or at least assert that this port needs to be open, so that the user can set it manually).

But I think this can be done in a separate PR.

File: pkgs/development/tools/unityhub/default.nix
Comment (Philipp-M) [inline pkgs/development/tools/unityhub/default.nix:89]: Maybe add `git` here for git based dependencies in the package manager? It could also be behind a feature-flag in the package arguments (but I don't think that's necessary).

**Comment by Philipp-M** [comment]
Just tested it, works great so far.

The xdg-scheme-handler is also working for me, as it's provided in the `.desktop` file.
@hyphenrf have you installed it correctly?

File: pkgs/development/tools/unityhub/default.nix
Comment (Philipp-M) [reply pkgs/development/tools/unityhub/default.nix]: AFAIK `dpkg-deb -x $src ` or `dpkg -x $src` is ok and the usual way to do this (results in the same).

**Comment by hyphenrf** [comment]
I've run it through a shell, so scheme handlers (expectedly) did not work for me, but the firewall problem was present no matter the way I installed it (shell or switch).

**Comment by huantianad** [comment]
Do we know if the there's a specific port that the app uses? Ideally we'd be able to just open that port.

Imo this feels like a bug either in the app or NixOS: the user shouldn't be required to disable or open their firewall to have full functionality of Unity Hub. Maybe we should file a report upstream?

**Comment by mainrs** [comment]
I do receive a crash when launching a new project based on the `3D (URP)` template:

```
Builtin Sky manager started.
Caught fatal signal - signo:11 code:128 errno:0 addr:(nil)
Obtained 28 stack frames.
#0  0x00558a97df3edc in burst_signal_handler(int, siginfo_t*, void*)
#1  0x007f6834791bf0 in __sigaction
#2  0x007f683bad16d9 in g_type_check_instance
#3  0x007f683bac5d74 in g_signal_handlers_disconnect_matched
#4  0x00558a9940c033 in DisconnectGtkSignal(void*, GtkCallbackResponse (*)(), void*)
#5  0x00558a98c90cfd in MenuItem::~MenuItem()
#6  0x00558a98c95673 in RemoveItemFromMenu(std::list<MenuItem, std::allocator<MenuItem> >*, core::basic_string_ref<char>)
#7  0x00558a98c956b3 in RemoveItemFromMenu(std::list<MenuItem, std::allocator<MenuItem> >*, core::basic_string_ref<char>)
#8  0x00558a98c954e3 in MenuController::RemoveMenuItem(core::basic_string_ref<char>)
#9  0x00558a978fe81c in Menu_CUSTOM_RemoveMenuItem(ScriptingBackendNativeStringPtrOpaque*)
#10 0x00000041a2f243 in (wrapper managed-to-native) UnityEditor.Menu:RemoveMenuItem (string)
#11 0x00000041a2e51f in UnityEditor.EditorApplication/<>c__DisplayClass110_0:<CallDelayed>b__0 ()
#12 0x007f6654046258 in mono_get_runtime_build_info
#13 0x007f66541e513a in mono_runtime_invoke
#14 0x007f66541e5084 in mono_runtime_invoke
#15 0x00558a98212903 in scripting_method_invoke(ScriptingMethodPtr, ScriptingObjectPtr, ScriptingArguments&, ScriptingExceptionPtr*, bool)
#16 0x00558a981f0a67 in ScriptingInvocation::Invoke(ScriptingExceptionPtr*, bool)
#17 0x00558a9831e50f in Scripting::UnityEditor::EditorApplicationProxy::Internal_InvokeTickEvents(ScriptingExceptionPtr*)
#18 0x00558a9936ba93 in Application::TickTimer()
#19 0x00558a9940b67e in MainMessageIteration(void*)
#20 0x007f6837d1a8e4 in g_main_context_dispatch
#21 0x007f6837d1ac88 in g_main_context_iterate.constprop.0
#22 0x007f6837d1af73 in g_main_loop_run
#23 0x007f6834e1e3c5 in gtk_main
#24 0x00558a9940a81c in main
#25 0x007f683477d24e in __libc_start_call_main
#26 0x007f683477d309 in __libc_start_main
#27 0x00558a97599029 in _start
```

Is something missing from the environment?

**Comment by huantianad** [comment]
@mainrs what Unity version is this template using?

**Comment by mainrs** [comment]
> @mainrs what Unity version is this template using?

2022.1.23f1 is what I have installed. However, it also crashes if I load a sample project.

**Comment by Philipp-M** [comment]
Looks like this: https://issuetracker.unity3d.com/issues/linux-editor-crashes-at-burst-signal-handler-when-opening-floating-windows

Use a recent Unity version. I think this is unrelated to this PR

**Comment by huantianad** [comment]
Yep, I also had the same crash before with older Unity versions and was fixed with an update.

**Comment by TheOddler** [comment]
This is the first time I'm trying to review a nixpkgs pr and not very good with Nix yet, so sorry if I'm missing something obvious.

Here's what I did and where I'm stuck:

- I cloned nixpkgs
- Ran  `nix run nixpkgs#nixpkgs-review pr 204005`
- In the nix-shell I then ran `unityhub`
- It starts correctly
- I click "sign in"
- It then tries to redirect me, I guess back to the hub application itself, but that doesn't seem to work

Is there something I can do to make the sign-in work from this nix-shell so I can test further?

**Review by TheOddler — APPROVED** [review]
From what I can see this works (see previous comment for more info).

**Comment by Philipp-M** [comment]
> * It then tries to redirect me, I guess back to the hub application itself, but that doesn't seem to work
> Is there something I can do to make the sign-in work from this nix-shell so I can test further?

I guess for that you may have to install it really, so that your host browser (or rather `xdg-open`) knows where to find the desktop file of unity-hub

>  It seems it picked up the unity installs I installed through the flatpak image. And opening a project with those worked.
I also installed a new version of Unity and created a new project with that, and that worked as well.

Yes Unity is installed in the home directory at `~/Unity` that's why the flatpak package and this can share the same package.

**Review by Philipp-M — APPROVED** [review]
IMHO I think we can merge that (I would still add the git dependency I've commented, to avoid weird issues when using the package manager).

The current issues with NixOS can IMO be ironed out in a follow-up PR (finding out the correct port, that is needed, or telling the user a warning to open the firewall when it's closed and there are some issues etc.). But as the current state is broken, I think it's better to get it working again (by merging this PR).

**Comment by mainrs** [comment]
It would be great if a new issue is created once this PR is merged to keep track for it. And to have a place to exchange and list the problems that still need a fix.

**Comment by huantianad** [comment]
@Philipp-M @mainrs sounds good, I'll open an issue to track this. 

Also just added `git` as a dependency, thanks!

**Comment by huantianad** [comment]
Just to make sure, there isn't anything else I need to do to get this PR done, adding the approvals label should make it noticeable to nixpkgs maintainers to consider for merge? I'm assuming it's just somewhere in the backlog hehe

**Comment by headblockhead** [comment]
> Perhaps @headblockhead, are you still having trouble with Unity 2022 and could this possibly be a NixOS and not unity bug?

Hello! Sorry for the late reply, I think that my issues are probably from Unity than NixOS as since upgrading to the latest LTS version of Unity, I have had no issues. Sorry for bringing it up!

If I find any other issues, I will check to make sure it is a NixOS problem first. :P

**Comment by huantianad** [comment]
Thanks for letting me know! Yeah it’s tricky to tell sometimes if an error is because of Unity or NixOS, no need to regret mentioning it. Better to be thorough!

File: pkgs/development/tools/unityhub/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/unityhub/default.nix:7]:
```suggestion
  pname = "unityhub";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
we want to use pname+version

**Pull request merged**
